plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -199,6 +221,55 @@ newtype ExBudgetMode cost (uni :: GHC.Type -> GHC.Type) fun = ExBudgetMode
     { unExBudgetMode :: forall s. ST s (ExBudgetInfo cost fun s)
     }
 
+{- Note [Cost slippage]
+Tracking the budget usage for every step in the machine adds a lot of overhead. To reduce this,
+we adopt a technique which allows some overshoot of the budget ("slippage"), but only a bounded
+amount.
+
+To do this we:
+- Track all the machine steps of all kinds in an optimized way in a 'WordArray'.
+- Actually "spend" the budget when we've done more than some fixed number of steps, or at the end.
+
```

Comment:
I think it'd be worth mentioning here that we're only allowing slippage for machine steps.  When we call a (potentially expensive) builtin we charge for it _right then_, so we're only being lax with the cheapish machine steps.

Reply:
Added a note.